app/common/lib/ledgerUtil.js

```diff
@@ -241,35 +263,184 @@ const getMediaData = (xhr, type) => {
   let duration = 0
+
+  if (data == null) {
+    return duration
+  }
+
   switch (type) {
-    case ledgerMediaProviders.YOUTUBE: {
-      duration = getYouTubeDuration(data)
-      break
-    }
+    case ledgerMediaProviders.YOUTUBE:
+      {
+        duration = getYouTubeDuration(data)
+        break
+      }
+    case ledgerMediaProviders.TWITCH:
+      {
+        duration = getTwitchDuration(state, data, mediaKey)
+        break
+      }
   }
 
   return duration
 }
 
+const generateMediaCacheData = (parsed, type) => {
+  let data = Immutable.Map()
+
+  if (parsed == null) {
+    return data
+  }
+
+  switch (type) {
+    case ledgerMediaProviders.TWITCH:
+      {
+        data = generateTwitchCacheData(parsed)
+        break
+      }
+  }
+
+  return data
+}
+
+const generateTwitchCacheData = (parsed) => {
+  if (parsed == null) {
+    return Immutable.Map()
+  }
+
+  if (parsed.properties) {
+    return Immutable.fromJS({
+      event: parsed.event,
+      time: parsed.properties.time
+    })
+  }
+
+  return Immutable.fromJS({
+    event: parsed.event
+  })
+}
+
+const getMediaFavicon = (providerName) => {
+  let image = null
+
+  if (!providerName) {
+    return image
+  }
+
+  providerName = providerName.toLowerCase()
+
+  switch (providerName) {
+    case ledgerMediaProviders.YOUTUBE:
+      {
+        image = require('../../../img/mediaProviders/youtube.png')
+        break
+      }
+    case ledgerMediaProviders.TWITCH:
+      {
+        image = require('../../../img/mediaProviders/twitch.svg')
+        break
+      }
+  }
+
+  return image
+}
+
+const getTwitchDuration = (state, data, mediaKey) => {
+  if (data == null || mediaKey == null) {
+    return 0
+  }
+
+  const previousData = ledgerVideoCache.getDataByVideoId(state, mediaKey)
+
+  if (previousData.isEmpty() && data.event === 'video-play') {
+    return milliseconds.second * 10
+  }
+
+  if (!data.properties) {
+    return 0
+  }
+
+  const oldMedia = ledgerVideoCache.getDataByVideoId(state, mediaKey)
+  let time = 0
+  const currentTime = parseFloat(data.properties.time)
+  const oldTime = parseFloat(previousData.get('time'))
+
+  if (
+    data.event === 'player_click_playpause' &&
+    oldMedia.get('event') !== 'player_click_playpause'
+  ) {
+    // User paused a video
+    time = currentTime - oldTime
+  } else if (previousData.get('event') === 'video-play') {
+    // From video play event to x event
+    time = currentTime - oldTime - 10
+  } else if (data.event === 'minute-watched') {
+    // Minute watched event
+    time = currentTime - oldTime
+  } else if (data.event === 'vod_seek' && oldMedia.get('event') !== 'player_click_playpause') {
+    // Vod seek event
+    time = currentTime - oldTime
+  }
+
+  if (isNaN(time)) {
+    return 0
+  }
+
+  if (time < 0) {
+    return 0
+  }
+
+  if (time > 120) {
+    time = 120 // 2 minutes
+  }
+
+  // we get seconds back, so we need to convert it into ms
+  time = time * 1000
```

Comment:
for readability, you can use `milliseconds.second` instead of 1000 😄

Reply:
done